Update Tor module to add option to encrypt Tor key

orbitalturtle opened the pull request:
This PR lays the groundwork for #6500, which adds the ability to encrypt a Tor private key on disk. We originally had this all in one PR, but needed to split out this refactoring up because changes to the Tor module need to be added in another PR. 

Summary of Changes:

- Lays groundwork in Tor package for encrypting private key

tvolk131 commented on tor/onionfile/onion_file.go:
```diff
@@ -0,0 +1,135 @@
+	DecryptPayloadFromReader(io.Reader, []byte) ([]byte, error)
+}
+
+// OnionFile is a file-based implementation of the OnionStore interface that
```
nit: `OnionFile` -> `File`

tvolk131 commented on tor/onionfile/onion_file_test.go:
```diff
@@ -0,0 +1,97 @@
+)
+
+var (
+	privateKey = []byte("RSA1024 hide_me_plz")
```
Any reason this isn't just `hide_me_plz` like it was before?

orbitalturtle replied:
Ah yup so that's added as a way to tell if the key is encrypted or not, see this line https://github.com/lightningnetwork/lnd/pull/6526/files#diff-a219f7652057175a72ef2276ada3396923cfd0c3e82b4149bd1705cb2a70a9d5R111

tvolk131 replied:
Gotcha, makes sense!

tvolk131 commented on tor/onionfile/onion_file_test.go:
```diff
@@ -0,0 +1,97 @@
+	return privateKey, nil
+}
+
+// TestOnionFile tests that the OnionFile implementation of the OnionStore
```
nit: Change `OnionFile` to either `onionFile` to match the package name, or `File` to match the new struct name.

tvolk131 reviewed (changes requested):
Looks good! Just a few nitpicky things that need to be cleaned up.

orbitalturtle commented:
Great, thanks for the review @tvolk131 ! Made those changes.

lightninglabs-deploy commented:
@orbitalturtle, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@orbitalturtle, remember to re-request review from reviewers when ready

tvolk131 commented on tor/onionfile/onion_file.go:
```diff
@@ -0,0 +1,135 @@
+// not exist, then ErrNoPrivateKey is returned.
+func (f *File) PrivateKey(_ OnionType) ([]byte, error) {
+	// Try to read the Tor private key to pass into the AddOnion call
+	if _, err := os.Stat(f.privateKeyPath); !errors.Is(err, os.ErrNotExist) {
```
I think this chunk can be made a bit easier to read:

```
if _, err := os.Stat(f.privateKeyPath); errors.Is(err, os.ErrNotExist) {
	return nil, ErrNoPrivateKey
}

privateKeyContent, err := ioutil.ReadFile(f.privateKeyPath)
if err != nil {
	return nil, err
}

// If the privateKey doesn't start with either v2 or v3 key params
// it's likely encrypted.
isEncrypted := !bytes.HasPrefix(privateKeyContent, []byte(V2KeyParam)) &&
	!bytes.HasPrefix(privateKeyContent, []byte(V3KeyParam))

if isEncrypted {
	// If the privateKeyContent is encrypted but --tor.encryptkey
	// wasn't set we return an error
	if !f.encryptKey {
		return nil, ErrEncryptedTorPrivateKey
	}
	// Attempt to decrypt the key
	reader := bytes.NewReader(privateKeyContent)
	privateKeyContent, err = f.encrypter.DecryptPayloadFromReader(reader, f.keyRing)
	if err != nil {
		return nil, err
	}
}
return privateKeyContent, nil
```

Crypt-iQ replied:
nit: can avoid indenting this block by exiting early if the file isn't found?

tvolk131 commented:
LGTM! I left one comment about a potential cleanup.

orbitalturtle commented:
Rebased this just FYI :)

Crypt-iQ commented on tor/onionfile/onion_file.go:
```diff
@@ -0,0 +1,135 @@
+		// If the privateKey doesn't start with either v2 or v3 key params
+		// it's likely encrypted.
+		if !bytes.HasPrefix(privateKeyContent, []byte(V2KeyParam)) &&
+			!bytes.HasPrefix(privateKeyContent, []byte(V3KeyParam)) {
```
not sure how the encryption is done, but is it possible that somebody is unlucky and their encrypted key matches the prefix?

Crypt-iQ replied:
i think this is like a 1/2^56 chance so i think it's ok

guggero replied:
The key params are strings of 5 or more characters, very very unlikely to ever get that from binary data at random.

Crypt-iQ commented on tor/cmd_onion.go:
```diff
@@ -3,88 +3,34 @@ package tor
-func (f *OnionFile) DeletePrivateKey(_ OnionType) error {
-	return os.Remove(f.privateKeyPath)
-}
+var _ OnionStore = (*onionfile.File)(nil)
```
It is a little odd that we're checking that `onionfile.File` (which is in a different package) adheres to this package's `OnionStore` interface. Normally they would live in the same package.

Crypt-iQ replied:
I believe the `OnionStore` iface can actually be unexported, but the methods exported. This lets people know that it can't be used outside of the `tor` package. This is what we do in the funding manager (https://github.com/lightningnetwork/lnd/blob/cf9a9864cf253dbbcac5904d360bbbde763e1ebe/funding/manager.go#L476)

orbitalturtle replied:
Makes sense!

Crypt-iQ commented on tor/onionfile/onion_file_test.go:
```diff
@@ -0,0 +1,95 @@
+
+	tempDir, err := ioutil.TempDir("", "onion_store")
+	require.NoError(t, err, "unable to create temp dir")
+
```
need `defer os.RemoveAll(tempDir)`

guggero replied:
Instead, we should start using `t.TempDir()` everywhere in tests, that automatically removes the directories.

Crypt-iQ reviewed (changes requested):
Looking at the PR that depends on this, it doesn't seem like there's a need for the `onionfile` pkg. I think this is mostly there

guggero commented on tor/cmd_onion.go:
```diff
@@ -142,11 +88,13 @@ func (c *Controller) prepareKeyparam(cfg AddOnionConfig) (string, error) {
 // prepareAddOnion constructs a cmd command string based on the specified
 // config.
-func (c *Controller) prepareAddOnion(cfg AddOnionConfig) (string, error) {
+func (c *Controller) prepareAddOnion(cfg AddOnionConfig) (
```
nit: break between parameters or return types, not before parenthesis.

guggero commented on tor/onionfile/onion_file.go:
```diff
@@ -0,0 +1,135 @@
+// NewOnionFile creates a file-based implementation of the OnionStore interface
+// to store an onion service's private key.
+func NewOnionFile(privateKeyPath string, privateKeyPerm os.FileMode,
+	encryptKey bool, encrypter EncrypterDecrypter, keyRing []byte,
```
nit: break between parameters, not before parenthesis.

guggero commented on tor/onionfile/onion_file.go:
```diff
@@ -0,0 +1,135 @@
+// encrypts the key before storing it if requested.
+func (f *File) StorePrivateKey(_ OnionType, privateKey []byte) error {
+	var b bytes.Buffer
+	var privateKeyContent []byte
```
set `privateKeyContent := privateKey` and move both `b` and `payload` into the `if`?
Then we don't need an `else` statement.

guggero commented on tor/onionfile/onion_file.go:
```diff
@@ -0,0 +1,135 @@
+		// it's likely encrypted.
+		if !bytes.HasPrefix(privateKeyContent, []byte(V2KeyParam)) &&
+			!bytes.HasPrefix(privateKeyContent, []byte(V3KeyParam)) {
+			// If the privateKeyContent is encrypted but --tor.encryptkey
```
nit: add empty line after multi-line `if` statement.
Also, multiple lines that are too long (use 8 spaces for tab character) and multiple comments without punctuation.

guggero commented on tor/onionfile/onion_file.go:
```diff
@@ -0,0 +1,135 @@
+
+		// If the privateKey doesn't start with either v2 or v3 key params
+		// it's likely encrypted.
+		if !bytes.HasPrefix(privateKeyContent, []byte(V2KeyParam)) &&
```
Invert the `if` condition and return early to avoid indentation?

guggero commented on tor/cmd_onion.go:
```diff
@@ -3,88 +3,34 @@ package tor
 	// StorePrivateKey stores the private key according to the
 	// implementation of the OnionStore interface.
-	StorePrivateKey(OnionType, []byte) error
+	StorePrivateKey(onionfile.OnionType, []byte) error
```
If the onion type is never actually used in these methods (and we're going to remove V2 soon anyway), should we remove them in this refactor?

orbitalturtle replied:
Yeah if they're not used anyway, makes sense to me

orbitalturtle replied:
Took a stab at removing these

guggero commented on tor/onionfile/onion_file.go:
```diff
@@ -0,0 +1,135 @@
+	var privateKeyContent []byte
+	payload := bytes.NewBuffer(privateKey)
+
+	if f.encryptKey {
```
This commit does more than just moving code around. I think we should adjust the commit message accordingly.
I think for this amount of code it's okay to keep it in one commit, but normally we'd follow this pattern:
 1. A commit that purely moves the code and updates other components to use the moved code (e.g. updates to package names)
 2. A commit that removes outdated functionality (e.g. the `OnionType`)
 3. A commit that adds new functionality

orbitalturtle replied:
Makes sense. I added these details to the commit message, but will keep this in mind in the future

guggero commented on tor/onionfile/onion_file_test.go:
```diff
@@ -0,0 +1,95 @@
+		privateKeyPath, 0600, false, mockEncrypter,
+		make([]byte, 0),
+	)
+	if _, err := onionFile.PrivateKey(V2); err != ErrNoPrivateKey {
```
Use `require.ErrorIs()` here and `require` everywhere else instead of `t.Fatalf()`.

guggero commented on tor/onionfile/onion_file_test.go:
```diff
@@ -0,0 +1,95 @@
+
+	storedPrivateKey, err := encryptedOnionFile.PrivateKey(V2)
+	require.NoError(t, err, "unable to retrieve encrypted private key")
+	if !bytes.Equal(storedPrivateKey, privateKey) {
```
Can use `require.Equal()`.

guggero reviewed (commented):
I know this PR (or its predecessor) is quite old, so we changed a few things along the line that require an update in the code.
But the main functionality looks pretty good, so there shouldn't be any major changes necessary and we should probably be able to get this in soon.

orbitalturtle commented:
Thanks for the reviews @Crypt-iQ and @guggero! I made those changes as well as rebased the PR.

Crypt-iQ commented on tor/cmd_onion.go:
```diff
@@ -3,88 +3,30 @@ package tor
 
 // AddOnionConfig houses all of the required parameters in order to
 // successfully create a new onion service or restore an existing one.
 type AddOnionConfig struct {
```
You should be able to remove the exported `OnionStore` now and replace it with `onionStore`. If I'm wrong then you can just get rid of the unexported `onionStore`

Crypt-iQ reviewed (commented):
Almost there. I don't see the point of the `onionfile` directory - am I missing something? It's not a package so I think that anything that imports it also imports the `tor` package so there wouldn't be a separation b/t the `tor` and `tor/onionfile`.

orbitalturtle commented:
@Crypt-iQ Ah so the original intent of moving onionfile comes from this conversation from way back when: https://github.com/lightningnetwork/lnd/pull/4458#discussion_r620729177
But perhaps since the Tor directory has evolved into being a module instead of a package, this no longer makes sense? Will defer to you and @guggero on that one.

guggero commented:
> @Crypt-iQ Ah so the original intent of moving onionfile comes from this conversation from way back when: [#4458 (comment)](https://github.com/lightningnetwork/lnd/pull/4458#discussion_r620729177) But perhaps since the Tor directory has evolved into being a module instead of a package, this no longer makes sense? Will defer to you and @guggero on that one.

Ah, I was trying to remember why we decided to move the file in the first place. I knew there was a reason but couldn't recall. I think the `EncrypterDecrypter` interface decouples things sufficiently so we don't have a dependency to `lnencrypt`. Therefore we can probably revert the move. Sorry about the back and forth...

orbitalturtle commented:
@guggero Ahh that's right, I forgot we had completely removed lnencrypt as a dependency. No worries! Updated the code to remove the new onionfile package and I updated the PR description, so this is ready for another look.

Crypt-iQ commented on tor/cmd_onion_test.go:
```diff
@@ -184,20 +203,30 @@ type mockStore struct {
+}
+
+func (m MockEncrypter) DecryptPayloadFromReader(_ io.Reader) ([]byte, error) {
+
```
nit: can remove newline

Crypt-iQ reviewed (approved):
lgtm pending nit fix

guggero commented on tor/cmd_onion.go:
```diff
@@ -52,31 +78,86 @@ var _ OnionStore = (*OnionFile)(nil)
+	if !f.encryptKey {
+		return nil, ErrEncryptedTorPrivateKey
+	}
+	// Attempt to decrypt the key.
```
nit: add empty line before comment.

guggero commented on tor/cmd_onion.go:
```diff
@@ -234,11 +317,18 @@ func (c *Controller) AddOnion(cfg AddOnionConfig) (*OnionAddr, error) {
-	// recreated later on.
-	if privateKey, ok := replyParams["PrivateKey"]; cfg.Store != nil && ok {
-		err := cfg.Store.StorePrivateKey(cfg.Type, []byte(privateKey))
+	// If a new onion service was created, use the new private key for storage
```
nit: missing punctuation.

guggero commented on tor/cmd_onion_test.go:
```diff
@@ -3,47 +3,66 @@ package tor
+	privateKey = []byte("RSA1024 hide_me_plz")
+)
+
+// TestOnionFile tests that the File implementation of the OnionStore
```
nit: missing punctuation.

guggero commented on tor/cmd_onion.go:
```diff
@@ -24,26 +33,43 @@ const (
+// EncrypterDecrypter is used for encrypting and decrypting the onion service
+// private key.
+type EncrypterDecrypter interface {
+	EncryptPayloadToWriter(bytes.Buffer, io.Writer) error
```
I think instead of `bytes.Buffer` this should just be `[]byte`.

orbitalturtle replied:
Yeah that makes sense as we need to convert it to bytes in the current function implementation anyway. I'll update this here and in the other PR as well

guggero commented on tor/cmd_onion_test.go:
```diff
@@ -184,20 +203,30 @@ type mockStore struct {
+
+func (m MockEncrypter) DecryptPayloadFromReader(_ io.Reader) ([]byte, error) {
+
+	return privateKey, nil
```
Should we return another string here so we can test that the encrypter was actually called?

orbitalturtle replied:
Another string?

guggero replied:
Something like `this_is_the_encrypted_key` instead of the normal private key. Then in the test make sure the "encrypted" key is returned instead of the plaintext `privateKey`.

orbitalturtle replied:
Ahhh I see. That makes sense. I feel iffy about calling it an encrypted key since it's being returned by the decrypter, so I made one called "anotherKey." Otherwise, does the new change look like what you had in mind?

guggero commented on tor/onionaddr.go:
```diff
@@ -45,6 +45,9 @@ type OnionAddr struct {
 	// Port is the port of the onion address.
 	Port int
+
+	// PrivateKey is the onion address' private key. Only returned if requested.
```
nit: this comment is no longer accurate, we're now always returning the private key.

guggero reviewed (commented):
Thank you for the updates!
A few last small comments, after those we're good to merge this :tada:

guggero reviewed (approved):
LGTM :tada:

guggero commented:
Pushed a new tag `tor/v1.1.0` with the merge commit.
